### ISSUE TITLE
dead_code false positive on matched enum

### ISSUE THREAD

**Issue by durka** [rust-lang/rust#29064]
[This code](http://is.gd/17G7yv) produces a dead code warning for the entire enum, but it isn't dead because the match expression refers to it. Or do I have the wrong definition of "dead"? (I generally assume when rustc says "this is dead" it means "you can comment this out".)

```rust
enum Color {
    Brown,
    Green,
}

fn main() {
    let x = None;
    match x {
        Some(t) => match t {
            Color::Brown => println!("it's brown"),
            Color::Green => println!("it's green"),
        },
        None => println!("nothing"),
    }
}
```

**Comment by sanxiyn** [comment]
Warning is intentional, as value of enum is never constructed and match arms can't be reached. It actually is the case that "you can comment this out".

**Comment by durka** [comment]
If you comment out the enum, you have to comment out the match as well, otherwise it doesn't compile. So it doesn't seem very dead, to me.

**Comment by sanxiyn** [comment]
See also #27559.

**Comment by arielb1** [comment]
@durka 

It is this way intentionally, exactly to catch variants that are not created anywhere.

**Comment by durka** [comment]
Ok, thanks for the explanation! At first I thought "but what if someone _else_ instantiates one?" but if you make it `pub` the misleading message goes away, so it's unlikely to affect real code.

**Comment by apoelstra** [comment]
Here is real code this affects. It's part of a macro-ization of the manual implementation of serde::Serialize. Grep for `impl<'a> serde::ser::MapVisitor for PointMapVisitor<'a>` to find the relevant code, which uses an integer to act as a state machine. In macros you cannot get a list of consecutive numbers like this, so I used an enum as:
```rust
            #[repr(u16)]
            #[derive(Copy, Clone)]
            #[allow(non_camel_case_types)]
            pub enum State { $($fe),* , Finished }

            pub struct MapVisitor<'a> {
                pub value: &'a super::$name,
                pub state: State,
            }

            impl<'a> ::serde::ser::MapVisitor for MapVisitor<'a> {
                fn visit<S>(&mut self, serializer: &mut S) -> Result<Option<()>, S::Error>
                    where S: ::serde::Serializer
                {
                    match self.state {
                        $(State::$fe => {
                            self.state = unsafe { ::std::mem::transmute(self.state as u16 + 1) };
                            Ok(Some(try!(serializer.visit_struct_elt(stringify!($fe), &self.value.$fe))))
                        })*
                        State::Finished => {
                            Ok(None)
                        }
                    }
                }
            }
        }
```
Later the enum is instantiated with a transmutation of 0.

The error still occurs, even though every single enum variant actually gets used.

**Comment by arielb1** [comment]
> Later the enum is instantiated with a transmutation of 0.

That's your problem. Just use `allow(dead_code)`.
